iree/turbine/kernel/wave/utils/general_utils.py

```diff
@@ -492,3 +497,61 @@ def infer_dim(expr):
         return expr
     dim_symbol = list(expr.free_symbols)[0]
     return dim_symbol
+
+
+def flatten_list(input_list: list):
+    flattened_list, _ = pytree.tree_flatten(input_list)
+    return flattened_list
+
+
+def topological_sort_with_dependencies(
+    node_ordering_constraint: list[fx.Node],
+    nodes_to_reorder: list[fx.Node],
+    reorder_node_dependencies: list[fx.Node] = [],
+):
+    """
+    This function help add dependencies nodes that are not explicitly specified
+    inside the cluster.
+
+    node_ordering_constraint: A list of nodes that is a subset of the nodes in `nodes_to_reorder` that represent the
+                              ordering constraint of nodes we'd like to enforce.
+
+    nodes_to_reorder: An list of nodes that we'd like to reorder and enforce the ordering from `node_ordering_constraint onto.
+
+    reorder_node_dependencies: If `node_to_reorder` is not a "complete" list of ops in the graph, we would need to suplement the
+                               function with it's misssing dependencies/argument nodes.
+    """
+    schedule_weight = dict.fromkeys(reorder_node_dependencies, 0)
+    workqueue = deque(nodes_to_reorder)
+    non_solved_counter = 0
```

Comment:
What is the point of non-solved counter? Is it for cycle detection? Should we add a check to make sure we don't get stuck in the while loop if a cycle is present?